File: 46.全排列.cpp
/*
 * @lc app=leetcode.cn id=46 lang=cpp
 *
 * [46] 全排列
 */

// @lc code=start
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    // https://leetcode-cn.com/problems/permutations/solution/quan-pai-lie-by-leetcode-solution-2/
    void dfs(vector<vector<int>> &res, vector<int> output, int first, int len){
        if(first == len){
            res.emplace_back(output);
            return;
        }
        for(int i = first; i < len; i++){
            // 动态维护数组
            swap(output[i], output[first]);
            // 继续递归填下一个位置
            dfs(res, output, first + 1, len);
            // 回溯后撤销操作
            swap(output[i], output[first]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        dfs(res, nums, 0, (int)nums.size()); 
        return res;
    }
};
// @lc code=end

File: 46.全排列.java
//labuladong 回溯算法
class Solution {
    List<List<Integer>> res = new LinkedList<>();
    public List<List<Integer>> permute(int[] nums) {
        LinkedList<Integer> track = new LinkedList<>();
        backTrack(nums, track);
        return res;
    }
    public void backTrack(int[] nums, LinkedList<Integer> track){
        // 终止条件
        if(track.size() == nums.length){
            //注意传入res一个track的拷贝对象（当前数组的快照），而不能直接传入res，res之后会不断变化的
            res.add(new LinkedList(track));
            return ;
        }

        for(int i = 0; i < nums.length; i++){
            if(track.contains(nums[i]))
                continue;
            track.add(nums[i]);
            backTrack(nums, track);
            track.removeLast();
        }
    }
}